engine/runtime/src/main/java/org/enso/interpreter/runtime/state/ExecutionEnvironment.java

```diff
@@ -0,0 +1,77 @@
+package org.enso.interpreter.runtime.state;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import org.enso.interpreter.node.expression.builtin.runtime.Context;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ExecutionEnvironment {
+
+  private final String name;
+  private final Map<String, Boolean> permissions;
+
+  public static final ExecutionEnvironment LIVE = initLive();
+  public static final ExecutionEnvironment DESIGN = new ExecutionEnvironment("design");
+
+  private static final ExecutionEnvironment initLive() {
+    Map<String, Boolean> contexts = new HashMap<>();
+    contexts.put("Input", true);
+    contexts.put("Output", true);
+    return new ExecutionEnvironment("live", contexts);
+  }
+
+  public ExecutionEnvironment(String name) {
+    this.name = name;
+    this.permissions = new HashMap<>();
+  }
+
+  private ExecutionEnvironment(String name, Map<String, Boolean> permissions) {
+    this.name = name;
+    this.permissions = permissions;
+  }
+
+  public String getName() {
+    return this.name;
+  }
+
+  @CompilerDirectives.TruffleBoundary
+  public ExecutionEnvironment withContextEnabled(Atom context) {
```

Comment:
Shouldn't we strive for making this function working on fast path?

Reply:
I would have to refactor pretty much any call to map i.e. almost every line in that method, to make native image build happy. I would say that kind of defeats the purpose?

Reply:
What's your purpose? Is the purpose is to run Enso fast? Well, in such case that's not happening as soon as somebody uses `withContextEnabled` - that's always going to be slow.

> refactor pretty much any call to map

Possibly using `HashMap` for storage of two values (`Input`, `Output`) wasn't good idea to begin with?

Reply:
Fine, replaced by two arrays that emulate the map. The reason why I don't hardcode those in fields is that I'm trying to be generic enough in preparation for potential future `Context` values.

Reply:
It could also be a big overkill given that **currently** we only have `Input` and `Output`. So if you feel strongly about it, I could just go back to storing those in fields.

Reply:
I am assuming that `withContextEnabled`/`withContextDisabled` operations are going to be spread all around the codebase possibly included inside of `.map` lambdas processing large data structures and we *want to make them fast* in order to avoid slowing down the rest of the interpreter. Is such assumption a false expectation? As I am seeing a disagreement and I don't understand what are its roots...

> replaced by two arrays that emulate the map

The code is now less readable, but at least it doesn't need to be behind `@TruffleBoundary`. However I cannot judge how well it partially evaluates without looking at IGV graphs. Btw. what do we want to see in IGV when we look at `withContextEnabled`?
- when already enabled - a single field/bit check to verify the context is already enabled
- when disabled - more work - field write and switching the `State` to a new configuration (possibly also a costly allocation of the new configuration) 

> trying to be generic enough in preparation for potential future Context values.

Generic code can't be as fast as code that handles/speculates on fixed set of values. 

> ... if **you feel** strongly ...

Maybe we could change one of our benchmarks to use `withContextEnabled` few million of times and compare it to a benchmark without context manipulation? Then **we all** would have something measurable to know how to **feel** about this matter.

Reply:
I will file a separate ticket to investigate if this becomes a perf issue.